extensions/sha256/circuit/src/sha256_chip/trace.rs

```diff
@@ -1,44 +1,118 @@
 use std::{
     array,
     borrow::{Borrow, BorrowMut},
+    cmp::min,
+    ptr::slice_from_raw_parts,
 };
 
 use openvm_circuit::{
-    arch::{Result, TraceStep, VmStateMut},
-    system::memory::{online::TracingMemory, MemoryAuxColsFactory},
+    arch::{
+        get_record_from_slice, CustomBorrow, MultiRowLayout, RecordArena, Result, TraceFiller,
+        TraceStep, VmStateMut,
+    },
+    system::memory::{
+        offline_checker::{MemoryReadAuxRecord, MemoryWriteAuxRecord},
+        online::TracingMemory,
+        MemoryAuxColsFactory,
+    },
 };
+use openvm_circuit_primitives::AlignedBytesBorrow;
 use openvm_instructions::{
     instruction::Instruction,
     program::DEFAULT_PC_STEP,
     riscv::{RV32_CELL_BITS, RV32_MEMORY_AS, RV32_REGISTER_AS, RV32_REGISTER_NUM_LIMBS},
     LocalOpcode,
 };
-use openvm_rv32im_circuit::adapters::{memory_read, tracing_read, tracing_write};
+use openvm_rv32im_circuit::adapters::{read_rv32_register, tracing_read, tracing_write};
 use openvm_sha256_air::{
-    get_flag_pt_array, u32_into_u16s, Sha256StepHelper, SHA256_BLOCK_BITS, SHA256_BLOCK_WORDS,
-    SHA256_H, SHA256_ROWS_PER_BLOCK, SHA256_WORD_U8S,
+    get_flag_pt_array, Sha256StepHelper, SHA256_BLOCK_BITS, SHA256_H, SHA256_ROWS_PER_BLOCK,
 };
 use openvm_sha256_transpiler::Rv32Sha256Opcode;
-use openvm_stark_backend::{p3_field::PrimeField32, p3_maybe_rayon::prelude::*};
+use openvm_stark_backend::{
+    p3_field::PrimeField32,
+    p3_matrix::{dense::RowMajorMatrix, Matrix},
+    p3_maybe_rayon::prelude::*,
+};
+use sha2::{Digest, Sha256};
 
 use super::{
     Sha256VmDigestCols, Sha256VmRoundCols, Sha256VmStep, SHA256VM_CONTROL_WIDTH,
     SHA256VM_DIGEST_WIDTH,
 };
 use crate::{
-    sha256_chip::{PaddingFlags, SHA256_READ_SIZE},
-    SHA256VM_ROUND_WIDTH, SHA256_BLOCK_CELLS,
+    sha256_chip::{PaddingFlags, SHA256_READ_SIZE, SHA256_REGISTER_READS, SHA256_WRITE_SIZE},
+    Sha256VmControlCols, SHA256VM_ROUND_WIDTH, SHA256VM_WIDTH, SHA256_BLOCK_CELLS,
+    SHA256_MAX_MESSAGE_LEN, SHA256_NUM_READ_ROWS,
 };
 
+#[derive(Clone, Copy)]
+pub struct Sha256VmMetadata {
+    pub num_blocks: u32,
+}
+
+#[repr(C)]
+#[derive(AlignedBytesBorrow, Debug, Clone)]
+pub struct Sha256VmRecord {
+    pub from_pc: u32,
+    pub timestamp: u32,
+    pub rd_ptr: u32,
+    pub rs1_ptr: u32,
+    pub rs2_ptr: u32,
+    pub dst_ptr: u32,
+    pub src_ptr: u32,
+    pub len: u32,
+
+    pub register_reads_aux: [MemoryReadAuxRecord; SHA256_REGISTER_READS],
+    pub write_aux: MemoryWriteAuxRecord<SHA256_WRITE_SIZE>,
+}
+
+pub struct Sha256VmRecordMut<'a> {
+    pub inner: &'a mut Sha256VmRecord,
+    // Having a continuous slice of the input is useful for fast hashing in `execute`
+    pub input: &'a mut [u8],
```

Comment:
shouldn't you be able to get this slice from memory? technically read_aux will have it all already so it's wasteful

Reply:
maybe a technicality because memory could be paged

Reply:
putting it here still feels weird - maybe switch back to making a `Vec`, or keep a scratch `Vec` inside the Step itself, and after @pjabbarzade 's PR is merged, we can try to switch to using pure slice